### PR TITLE
adds DataAccess project to Complex Service

### PR DESCRIPTION
**Pull request by markcmoore**

### CODE REVIEW

File: complex/.azure/azureup.yaml [hunk 1]
```diff
@@ -8,14 +8,15 @@ stages:
     jobs:
```
Comment (gpyofavv) [inline complex/.azure/azureup.yaml]: You shouldn't change this file! Just copy/paste it over from master.

File: complex/src/Revature.Complex.Api/Revature.Complex.Api.csproj [hunk 2]
```diff
@@ -7,8 +7,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.0.0" />
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.Api/Revature.Complex.Api.csproj]: Go ahead and sync this up: just copy and paste the red lines of code over your code. clean, build, test just to make sure.

File: complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+using System;
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs]: Remove unnecessary using statements in all code

File: complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+    public ComplexDbContext(DbContextOptions<ComplexDbContext> options)
+    : base(options) { }
+
+    public virtual DbSet<Complex> Complex { get; set; }
+
+    public virtual DbSet<AmenityComplex> AmenityComplex { get; set; }
+    public virtual DbSet<AmenityRoom> AmenityRoom { get; set; }
+    public virtual DbSet<Amenity> Amenity { get; set; }
+
+    public Guid cId1 = Guid.NewGuid();
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs]: You can also copy/paste a guid in and use guid.parse() to make it a guid: this way the guids are the same everytime. You're call, you could wait to make that change until you actually need it that way.

File: complex/src/Revature.Complex.DataAccess/Mapper.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Revature.Complex.Lib.Interface;
+using Logic = Revature.Complex.Lib.Models;
+using Entity = Revature.Complex.DataAccess.Entities;
+
+namespace Revature.Complex.DataAccess
+{
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.DataAccess/Mapper.cs]: documentation

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    /// <param name="lComplex"></param>
+    /// <returns></returns>
+    public async Task<string> CreateComplexAsync(Logic.Complex lComplex)
+    {
+      Entity.Complex complex = _map.MapComplextoE(lComplex);
+
+      await _context.AddAsync(complex);
+      await _context.SaveChangesAsync().ConfigureAwait(false);
+
+      return "Create!!";
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: Using a bool would be more professional.

File: complex/src/Revature.Complex.Lib/Models/Amenity.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Revature.Complex.Lib.Models
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.Lib/Models/Amenity.cs]: Document class, not necessarily every field and property.

Comment (gpyofavv) [reply complex/src/Revature.Complex.Lib/Models/Amenity.cs]: Same for all models

File: complex/src/Revature.Complex.Tests/Revature.Complex.Tests.csproj [hunk 8]
```diff
@@ -7,10 +7,16 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
-    <PackageReference Include="xunit" Version="2.4.0" />
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.Tests/Revature.Complex.Tests.csproj]: Same as comment on other cs proj: your versions should match his. Don;t bother playing with NuGet packages, just copy and paste.

File: housing/src/app/guards/auth.guard.spec.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
+import { TestBed, async, inject, getTestBed } from '@angular/core/testing';
```
Comment (gpyofavv) [inline housing/src/app/guards/auth.guard.spec.ts]: This doesn't belong in this pr

File: housing/src/app/guards/auth.guard.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
+// Auth0 Quickstart code taken from https://auth0.com/docs/quickstart/spa/angular2
```
Comment (gpyofavv) [inline housing/src/app/guards/auth.guard.ts]: Wrong pr

File: complex/.azure/azureup.yaml (hunk 1 above)
Comment (markcmoore) [reply complex/.azure/azureup.yaml]: imported from master for this PR

File: complex/src/Revature.Complex.Api/Revature.Complex.Api.csproj (hunk 2 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.Api/Revature.Complex.Api.csproj]: done!

File: complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs (hunk 4 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs]: noted. I'll wait, I think!

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 6 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: I changed the message. Will change to bool in another iteration...time permitting.

File: complex/src/Revature.Complex.Tests/Revature.Complex.Tests.csproj (hunk 8 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.Tests/Revature.Complex.Tests.csproj]: copied!

File: housing/src/app/guards/auth.guard.spec.ts (hunk 9 above)
Comment (markcmoore) [reply housing/src/app/guards/auth.guard.spec.ts]: Replaced the housing with the master version.

File: housing/src/app/guards/auth.guard.ts (hunk 10 above)
Comment (markcmoore) [reply housing/src/app/guards/auth.guard.ts]: replaced

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+        await _context.SaveChangesAsync();
+
+        return "Amenity deleted";
+      }
+      catch (InvalidOperationException ex)
+      {
+        log.LogWarning($"{ex}: Unable to delete the amenity.");
+        return "cannot find target amenity to delete";
+      }
+
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: Remove empty lines so
    }
  }
}

Comment (gpyofavv) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: not 
}

}

}

File: complex/src/Revature.Complex.Lib/Models/Complex.cs [hunk 12]
```diff
@@ -0,0 +1,16 @@
+using System;
```
Comment (gpyofavv) [inline complex/src/Revature.Complex.Lib/Models/Complex.cs]: This may cause to issues once you've gotten your other PR into master.

File: housing/.azure/azureup.yaml [hunk 13]
```diff
@@ -8,7 +8,8 @@ stages:
     jobs:
```
Comment (gpyofavv) [inline housing/.azure/azureup.yaml]: Don't change this.

File: housing/auth0rules.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
+// These rules are currently in use at auth0.com and are listed here
```
Comment (gpyofavv) [inline housing/auth0rules.js]: Why is this in PR?

File: housing/karma.conf.js [hunk 15]
```diff
@@ -7,6 +7,7 @@ module.exports = function (config) {
     frameworks: ['jasmine', '@angular-devkit/build-angular'],
```
Comment (gpyofavv) [inline housing/karma.conf.js]: for all things in housing/, remove from PR

File: address/src/Revature.Address.Tests/Revature.Address.Tests.csproj [hunk 16]
```diff
@@ -7,10 +7,16 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
-    <PackageReference Include="xunit" Version="2.4.0" />
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply address/src/Revature.Address.Tests/Revature.Address.Tests.csproj]: restored... again...

File: address/.azure/azureup.yaml [hunk 17]
```diff
@@ -8,14 +8,15 @@ stages:
     jobs:
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply address/.azure/azureup.yaml]: restored... again

File: housing/.azure/azureup.yaml (hunk 13 above)
Comment (markcmoore) [reply housing/.azure/azureup.yaml]: it matches master now.

File: complex/src/Revature.Complex.DataAccess/Mapper.cs (hunk 5 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Mapper.cs]: done!

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 18]
```diff
@@ -0,0 +1,360 @@
...
+    /// Read single logic complex object from complex name and complex contact number
+    /// </summary>
+    /// <param name="name"></param>
+    /// <param name="phone"></param>
+    /// <returns></returns>
+    public async Task<Logic.Complex> ReadComplexByNameAndNumberAsync(string name, string phone)
+    {
+      try
+      {
+        List<Entity.Complex> complex = await _context.Complex.Where(c => c.ComplexName == name
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: done

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 11 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: done

File: housing/auth0rules.js (hunk 14 above)
Comment (markcmoore) [reply housing/auth0rules.js]: it's in master.

File: housing/karma.conf.js (hunk 15 above)
Comment (markcmoore) [reply housing/karma.conf.js]: this is in master

File: complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs (hunk 3 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs]: noted

File: address/src/Revature.Address.Api/Revature.Address.Api.csproj [hunk 19]
```diff
@@ -7,8 +7,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Design" Version="3.0.0" />
     <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks.EntityFrameworkCore" Version="3.0.0" />
-    <PackageReference Include="Serilog.AspNetCore" Version="3.0.0" />
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply address/src/Revature.Address.Api/Revature.Address.Api.csproj]: done

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 11 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: done

File: housing/auth0rules.js (hunk 14 above)
Comment (markcmoore) [reply housing/auth0rules.js]: its in master

File: housing/karma.conf.js (hunk 15 above)
Comment (markcmoore) [reply housing/karma.conf.js]: these things are in master.

**Comment by escalonn** [comment]
resolve conflicts and quality gate

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 20]
```diff
@@ -0,0 +1,358 @@
...
+    /// <returns></returns>
+    public async Task<List<Logic.Complex>> ReadComplexListAsync()
+    {
+      try
+      {
+        List<Entity.Complex> complices = await _context.Complex.ToListAsync();
+
+        return complices.Select(_map.MapEtoComplex).ToList();
+      }
+      catch (ArgumentNullException ex)
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: 1. why would argumentnullexception happen?
2. try-catch-throw with nothing else is just a waste of space.

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 21]
```diff
@@ -0,0 +1,358 @@
...
+
+    /// <summary>
+    /// Read single Logic complex object from complex Id
+    /// </summary>
+    /// <param name="complexId"></param>
+    /// <returns></returns>
+    public async Task<Logic.Complex> ReadComplexByIdAsync(Guid complexId)
+    {
+      Entity.Complex complexFind = await _context.Complex.FindAsync(complexId).ConfigureAwait(false);
+      return _map.MapEtoComplex(complexFind);
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: clarify behavior when entity is not found.

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 22]
```diff
@@ -0,0 +1,358 @@
...
+    /// <returns></returns>
+    public async Task<Logic.Complex> ReadComplexByNameAndNumberAsync(string name, string phone)
+    {
+      try
+      {
+        List<Entity.Complex> complex = await _context.Complex
+          .Where(c => c.ComplexName == name && c.ContactNumber == phone)
+          .AsNoTracking()
+          .ToListAsync();
+        return _map.MapEtoComplex(complex[0]);
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: just use `.First` if you are going to throw away any other matches anyway

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 23]
```diff
@@ -0,0 +1,358 @@
...
+      }
+      catch(Exception ex)
+      {
+        log.LogError($"{ex}: Cannot Find specific complex");
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: don't convert exceptions directly to strings. all log methods have an optional parameter for including an exception object with an error message for context like this string here.

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 24]
```diff
@@ -0,0 +1,358 @@
...
+    /// </summary>
+    /// <param name="update"></param>
+    /// <returns></returns>
+    public async Task<string> UpdateComplexAsync(Logic.Complex update)
+    {
+      Entity.Complex origin = _context.Complex.Find(update.ComplexId);
+
+      if (origin.ComplexName == null)
+      {
+        return "no such complex";
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: don't return any strings like this. this is not how we communicate between parts of our code. e.g. use a boolean return instead.

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs [hunk 25]
```diff
@@ -0,0 +1,358 @@
...
+      Entity.Complex target = _context.Complex.Find(complexId);
+
+      if (target.ComplexName == null)
+      {
+        return "no such complex";
+      }
+      else
+      {
+        _context.Remove(target);
+        await _context.SaveChangesAsync();
```
Comment (escalonn) [inline complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: be consistent with using or not using configureawait(false)

File: complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs [hunk 26]
```diff
@@ -0,0 +1,172 @@
...
+    public ComplexDbContext(DbContextOptions<ComplexDbContext> options)
+    : base(options) { }
+
+    public virtual DbSet<Complex> Complex { get; set; }
+
+    public virtual DbSet<AmenityComplex> AmenityComplex { get; set; }
+    public virtual DbSet<AmenityRoom> AmenityRoom { get; set; }
+    public virtual DbSet<Amenity> Amenity { get; set; }
+
+    private Guid cId1 = Guid.NewGuid();
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs]: use this => private Guid cId2 = Guid.Parse("[real guid here]");

[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs]: all instances changed.

File: complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs [hunk 27]
```diff
@@ -0,0 +1,28 @@
...
+{
+  public class Amenity
+  {
+    /// <summary>
+    /// Primary key of Amenity table
+    /// </summary>
+    public Guid AmenityId { get; set; }
+
+    /// <summary>
+    /// Amentity name cell of Amenity table
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/Amenity.cs]: corrected!

File: complex/src/Revature.Complex.DataAccess/Entities/Complex.cs [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+    /// Complex name cell of Complex table
+    /// </summary>
+    public string ComplexName { get; set; }
+
+    /// <summary>
+    /// Contact number cell of Complex table
+    /// </summary>
+    public string ContactNumber { get; set; }
+
+    public virtual ICollection<AmenityComplex> AmentityComplex { get; set; }
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/Complex.cs]: corrected.

File: complex/src/Revature.Complex.DataAccess/Entities/AmenityRoom.cs [hunk 29]
```diff
@@ -0,0 +1,26 @@
...
+
+    /// <summary>
+    /// Foreign key of Amenity table
+    /// </summary>
+    public Guid AmenityId { get; set; }
+
+    /// <summary>
+    /// Room Id cell of AmenityRoom table
+    /// </summary>
+    public Guid RoomId { get; set; }
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/AmenityRoom.cs]: This is a FK representing a room in the Room service. Corrected description in PR.

File: complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs [hunk 30]
```diff
@@ -0,0 +1,172 @@
...
+            ContactNumber = "3332221111"
+          }
+        );
+
+      });
+
+      modelBuilder.Entity<Amenity>(entity =>
+      {
+        entity.Property(e => e.AmenityId)
+            .IsRequired();
```
[the remark this answers is not in the document]
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Entities/ComplexDbContext.cs]: corrected

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 21 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: try/catch added

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 22 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: Changed to return just the first result.

File: complex/src/Revature.Complex.DataAccess/Repository/Repository.cs (hunk 23 above)
Comment (markcmoore) [reply complex/src/Revature.Complex.DataAccess/Repository/Repository.cs]: all 'ex' changed to 'ex.Message'
